### ISSUE TITLE
JVP softmax implementation is missing a stop_gradient, leading to training instability

### ISSUE THREAD

**Issue by AhmedOmran0** [google/jax#16093]
### Description

In [jax/_src/nn/functions.py](https://github.com/google/jax/tree/main/jax/_src/nn/functions.py) there's a new implementation of JVP-based softmax.

Recently my team has been struggling with exploding gradients on training Transformer models, which we didn't have before. After bisecting all changes in JAX and other libraries we use, we narrowed it down to this change. The new softmax implementation is missing a `stop_gradient` operator around the x_max evaluation, which was done on the deprecated softmax implementation, and is highlighted as a canonical use case: 

https://www.tensorflow.org/api_docs/python/tf/stop_gradient

### What jax/jaxlib version are you using?

jax v0.4.10

### Which accelerator(s) are you using?

TPU

### Additional system info

Linux

**Comment by mattjj** [comment]
Just adding some context we discussed in chat: after adding the `custom_jvp`, the `stop_gradient` in the `softmax` implementation is always a no-op, since we no longer autodiff through the `softmax` implementation. Just to be extra sure, I checked:

<img width="952" alt="image" src="https://github.com/google/jax/assets/1458824/e6cb6513-c974-4226-8550-720069c548f1">

The assert triggers if we remove the custom_jvp rule, but with the custom_jvp rule in place it doesn't.
